discord/automod.py

```diff
@@ -0,0 +1,420 @@
+
+        return {}
+
+
+class AutoModRule:
+    """Represents an auto moderation rule.
+
+    Attributes
+    -----------
+    id: :class:`int`
+        The ID of the rule.
+    guild: :class:`Guild`
+        The guild the rule is for.
+    name: :class:`str`
+        The name of the rule.
+    creator_id: :class:`int`
+        The ID of the user that created the rule.
+    trigger: :class:`AutoModTrigger`
+        The rule's trigger.
+    enabled: :class:`bool`
+        Whether the rule is enabled.
+    raw_exempt_roles: List[:class:`int`]
+        The IDs of the roles that are exempt from the rule.
+    raw_exempt_channels: List[:class:`int`]
+        The IDs of the channels that are exempt from the rule.
+
+    .. versionadded:: 2.0
+    """
+
+    __slots__ = (
+        '_state',
+        'id',
+        'guild',
+        'name',
+        'creator_id',
+        'event_type',
+        'trigger',
+        'enabled',
+        'raw_exempt_roles',
+        'raw_exempt_channels',
+        '_actions',
+    )
+
+    def __init__(self, *, data: AutoModerationRulePayload, guild: Guild, state: ConnectionState) -> None:
+        self._state: ConnectionState = state
+        self.guild: Guild = guild
+        self.id: int = int(data['id'])
+        self.name: str = data['name']
+        self.creator_id = int(data['creator_id'])
+        self.event_type: AutoModRuleEventType = try_enum(AutoModRuleEventType, data['event_type'])
+        self.trigger: AutoModTrigger = AutoModTrigger.from_data(data['trigger_type'], data=data.get('trigger_metadata'))
+        self.enabled: bool = data['enabled']
+        self.raw_exempt_roles: List[int] = [int(role_id) for role_id in data['exempt_roles']]
+        self.raw_exempt_channels: List[int] = [int(channel_id) for channel_id in data['exempt_channels']]
+        self._actions: List[AutoModerationActionPayload] = data['actions']
+
+    def __repr__(self) -> str:
+        return f'<AutoModRule id={self.id} name={self.name!r} guild={self.guild!r}>'
+
+    def to_dict(self) -> AutoModerationRulePayload:
+        ret: AutoModerationRulePayload = {
+            'id': str(self.id),
+            'guild_id': str(self.guild.id),
+            'name': self.name,
+            'creator_id': str(self.creator_id),
+            'event_type': self.event_type.value,
+            'trigger_type': self.trigger.type.value,
+            'trigger_metadata': self.trigger.to_metadata_dict(),
+            'actions': [action.to_dict() for action in self.actions],
+            'enabled': self.enabled,
+            'exempt_roles': [str(role_id) for role_id in self.raw_exempt_roles],
+            'exempt_channels': [str(channel_id) for channel_id in self.raw_exempt_channels],
+        }  # type: ignore # trigger types break the flow here.
+
+        return ret
+
+    @property
+    def creator(self) -> Optional[User]:
+        """Optional[:class:`User`]: The user that created this rule."""
+        return self._state.get_user(self.creator_id)
```

Comment:
Surely this should have an upgrade flow for Member considering there's a guild attached?